sympy/simplify/cse_main.py

```diff
@@ -40,6 +40,195 @@
+    {x0: x1/x2, x3: x0 - exp(x2)}
+
+    Parameters
+    ==========
+
+    replacements : ``List[Tuple[Symbol, Expr]]``
+        The symbol-expression replacements pairs resulting from the CSE.
+    reduced_exprs : ``List[Expr]``
+        The replaced expressions resulting from the CSE.
+    replacements_mapping : ``Dict[Symbol, Expr]``
+        A mapping of replacements symbols to their corresponding expression.
+
+
+    See Also
+    ========
+
+    cse: Function for common subexpression elimination.
+
+    """
+
+    @classmethod
+    def from_reduced_exprs(cls, reduced_exprs, replacements_mapping, *,
+                           as_list=True):
+        """Alternate constructor to instantiate a ``CseExpr`` from a list of
+        reduced expressions and a mapping of replacements.
+
+        Examples
+        ========
+
+        A ``CseExpr`` can be constructed from a reduced expression and a
+        dictionary of replacements mapping SymPy symbols to SymPy expressions:
+
+        >>> from sympy import symbols, sin, exp
+        >>> from sympy.simplify.cse_main import CseExpr
+        >>> x0, x1, x2, x3 = symbols('x0, x1, x2, x3')
+        >>> reduced_expr = x3 * (x3 + sin(x0))
+        >>> replacements_mapping = {x0: x1 / x2, x3: x0 - exp(x2)}
+        >>> cse_expr = CseExpr.from_reduced_exprs(reduced_expr,
+        ...                                       replacements_mapping)
+        >>> print(cse_expr)
+        ([(x0, x1/x2), (x3, x0 - exp(x2))], [x3*(x3 + sin(x0))])
+
+        As only a single expression is passed as an argument to
+        ``reduced_exprs``, this method will wrap it into a single-item ``list``.
+        Alternatively, the reduced expression can be kept unchanged using the
+        ``as_list`` parameter:
+
+        >>> cse_expr = CseExpr.from_reduced_exprs(reduced_expr,
+        ...                                       replacements_mapping,
+        ...                                       as_list=False)
+        >>> print(cse_expr)
+        ([(x0, x1/x2), (x3, x0 - exp(x2))], x3*(x3 + sin(x0)))
+
+        Parameters
+        ==========
+
+        reduced_exprs : ``Expr`` or ``List[Expr]``
+            The reduced expression(s).
+        replacements_mapping : ``Dict[Symbol, Expr]``
+            Pairs of replacements present in ``exprs``.
+        as_list : ``bool``, default is ``True``
+            Whether the argument passed to ``reduced_exprs`` should be converted
+            to a ``list``. With the default of ``True``, if an iterable is
+            passed to ``reduced_exprs`` then it is converted to a ``list``, or
+            if a single ``Expr`` is passed then it is wrapped into a single-item
+            ``list``. If ``False``, then the argument passed to
+            ``reduced_exprs`` is left as is, meaning that the value returned by
+            the ``reduced_exprs`` property is not necessarily a ``list``. This
+            parameter is similar to the ``list`` parameter of the ``cse``
+            function.
```

Comment:
I would generally prefer to have an alternate function rather than a parameter like this `as_list`. The method here is called `from_reduced_exprs` (plural) with the parameter `reduced_exprs` (plural) but yet `reduced_exprs` can be a list and then we end up needing to have `as_list` to say whether we want a single expression to be treated as an expression or a list. This seems like a messy signature to me.

Would it not make sense to have something like `from_reduced_expr` (singular) and `from_reduced_expr_list` (plural)?

Then if you want an Expr you call `from_reduced_expr` and if you want a list of length 1 you can `from_reduced_expr_list` with a list of length 1.

Reply:
Or alternatively why do we need `as_list` and why is it necessary to put a single Expr in a list when the caller can do that?

Reply:
Another alternative is that there could be separate types CseExpr and CseExprList.

Basically I don't think that it is a good idea to build this ambiguity about internal types into any new type like CseExpr. If there is some other code (e.g. lambdify) that needs this ambiguity then it would be better to localise the ambiguity there rather than making the CseExpr type inherently type-ambiguous.